src/main/kotlin/build/buf/intellij/resolve/BufRootsProvider.kt

```diff
@@ -14,57 +14,130 @@
+            }
+            val home = env["HOME"]
+            if (home != null) {
+                return Path.of(home, ".cache", "buf")
+            }
+            // TODO: LOCALAPPDATA on Windows
+            return Path.of(SystemProperties.getUserHome(), ".cache", "buf")
+        }
 
-        val bufCacheFolder: VirtualFile?
-            get() = LocalFileSystem.getInstance().findFileByNioFile(bufCacheFolderBase)
+        fun getBufCacheFolderV1(env: Map<String, String> = System.getenv()) : VirtualFile? {
+            return LocalFileSystem.getInstance().findFileByNioFile(getBufCacheFolderBase(env))
                 ?.findFileByRelativePath("v1/module/data")
+        }
+
+        fun getBufCacheFolderV2(env: Map<String, String> = System.getenv()) : VirtualFile? {
+            return LocalFileSystem.getInstance().findFileByNioFile(getBufCacheFolderBase(env))
+                ?.findFileByRelativePath("v2/module")
+        }
+
+        private fun findManifestV2(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): Manifest? {
+            val repoCacheFolder = getBufCacheFolderV2(env)
+                ?.findFileByRelativePath("${mod.remote}/${mod.owner}/${mod.repository}") ?: return null
+            return Manifest.fromCommit(repoCacheFolder, mod.commit)
+        }
 
-        fun findModuleCacheFolder(mod: BufModuleCoordinates): VirtualFile? = bufCacheFolder
-            ?.findChild(mod.remote)
-            ?.findChild(mod.owner)
-            ?.findChild(mod.repository)
-            ?.findChild(mod.commit)
+        fun findModuleCacheFolderV1(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): VirtualFile? = getBufCacheFolderV1(env)
+            ?.findFileByRelativePath("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
+
+        fun getOrCreateModuleCacheFolderV2(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): VirtualFile? {
+            val bufPath = Paths.get(PathManager.getSystemPath(), "buf")
+            if (!bufPath.isDirectory() && !bufPath.toFile().mkdirs()) {
+                return null
+            }
+            val commitPath = bufPath.resolve("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
+            if (commitPath.isDirectory()) {
+                return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(commitPath)
+            }
+            val manifest = findManifestV2(mod, env) ?: return null
+            for (path in manifest.getPaths()) {
+                val canonicalPath = manifest.getCanonicalPath(path) ?: return null
+                val modulePath = commitPath.resolve(path)
+                if (!modulePath.parent.isDirectory()) {
+                    Files.createDirectories(modulePath.parent)
+                }
+                Files.copy(Path.of(canonicalPath), modulePath)
+            }
+            return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(commitPath)
```

Comment:
This will copy the CAS on disk from a Buf CLI to a directory structure suitable for use with IntelliJ. I tried to use virtual file systems to avoid this extra copy but nothing I tried would work end to end.